tslearn/early_classification.py

```diff
@@ -0,0 +1,312 @@
+                    self.pyhatyck_["pyhatycks{0}".format(k)].append(
+                        conf_matrix
+                    )
+
+    def get_cluster_probas(self, X):
+        """
+        This function computes the probabilities of the time series xt to be in a cluster of the model
+
+        Parameters
+        ----------
+        X: vector
+            a time series that may be truncated
+
+        Returns
+        -------
+        vector : the probabilities of the given time series to be in the clusters of the model
+        """
+        diffs = X[np.newaxis, :] - self.cluster_.cluster_centers_[:, len(X)]
+        distances_clusters = np.linalg.norm(diffs, axis=2)
+        distances_clusters = np.asarray(distances_clusters)
+        minimum_distance = np.min(distances_clusters)
+        average_distance = np.mean(distances_clusters)
+        minimum_distance_to_average = (
+            average_distance - minimum_distance
+        ) / average_distance
+        delta = (average_distance - distances_clusters) / average_distance
+        probas = 1.0 / (
+            np.exp(self.lamb * minimum_distance_to_average)
+            + np.exp(self.lamb * (minimum_distance_to_average - delta))
+            + 1e-6
+        )
+        return probas / probas.sum()
+
+    def _expected_cost(self, X, tau):
+        """
+        From a incomplete series xt, this function compute the expected cost of a prediction made at time "last time of
+        xt + tau"
+
+        Parameters
+        ----------
+        X: vector
+            a time series that may be truncated for which we want to have the cost
+        tau: int
+            gives the future time for which the cost is calculated
+
+        Returns
+        --------
+        float : the computed cost
+        """
+        cost = 0
+        proba_clusters = self.get_cluster_probas(X=X)
+        truncated_t = X.shape[-1]
+        for k in range(0, self.n_clusters):
+            sum_pyhatyck = np.sum(self.pyhatyck_["pyhatycks{0}".format(k)][
+                                truncated_t + tau - self.minimum_time_stamp
+                            ], axis=0)
+            sum_global = np.dot(sum_pyhatyck, self.pyck_[:, k])
+            cost = cost + proba_clusters[:, k] * sum_global
+
+        return cost
+
+    def minimize_integer(self, end_of_time, xt):
+        """
+        We want to minimize a function "funct" according a time series "xt" for a list of integers from 0 to
+        "end_of_time"
+
+        Parameters
+        ----------
+        end_of_time: int
+            the highest integer at which the function to be minimized is calculated
+        function_to_minimize: funct
+            the function of interest
+        xt: vector
+            the time series related to the function "funct"
+
+        Returns
+         ------
+         int : The integer in {0,...,stop} that minimizes "funct"
+         float : the so fat minimum cost
+        """
+        tau_star = 0
```

Comment:
This function could probably be 3-4 lines of code (does not change efficiency, but seems more readable to me):

```python
costs = [self._expected_cost(xt, tau) for tau in range(end_of_time)]
tau_star = numpy.argmin(costs)
min_cost = costs[tau_star]
return tau_star, min_cost
```